sources/ubuntu-http.go

```diff
@@ -30,19 +32,263 @@ func NewUbuntuHTTP() *UbuntuHTTP {
+	}
+	defer unix.Unmount(imageDir, 0)
+
+	err = shared.RunCommand("rsync", "-qa", filepath.Join(imageDir, "system-data"), rootfsDir)
+	if err != nil {
+		return err
+	}
+
+	// Create all the needed paths and links
+
+	dirs := []string{"bin", "dev", "initrd", "lib", "mnt", "proc", "root", "sbin", "sys"}
+
+	for _, d := range dirs {
+		err := os.Mkdir(filepath.Join(rootfsDir, d), 0755)
+		if err != nil {
+			return err
+		}
+	}
+
+	links := []struct {
+		target string
+		link   string
+	}{
+		{
+			"lib",
+			filepath.Join(rootfsDir, "lib64"),
+		},
+		{
+			"/bin/busybox",
+			filepath.Join(rootfsDir, "bin", "sh"),
+		},
+		{
+			"/bin/init",
+			filepath.Join(rootfsDir, "sbin", "init"),
+		},
+	}
+
+	for _, l := range links {
+		err = os.Symlink(l.target, l.link)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Copy system binaries
+
+	binaries := []struct {
+		source string
+		target string
+	}{
+		{
+			getFullPath("busybox"),
+			filepath.Join(rootfsDir, "bin", "busybox"),
+		},
+		{
+			getFullPath("cpio"),
+			filepath.Join(rootfsDir, "bin", "cpio"),
+		},
+		{
+			getFullPath("mount.fuse"),
+			filepath.Join(rootfsDir, "bin", "mount.fuse"),
+		},
+		{
+			getFullPath("pivot_root"),
+			filepath.Join(rootfsDir, "bin", "pivot_root"),
+		},
+		{
+			getFullPath("squashfuse"),
+			filepath.Join(rootfsDir, "bin", "squashfuse"),
+		},
+	}
+
+	for _, b := range binaries {
+		err := lxd.FileCopy(b.source, b.target)
+		if err != nil {
+			return err
+		}
+
+		err = os.Chmod(b.target, 0755)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Copy needed libraries
 
-	if strings.ContainsAny(definition.Image.Release, "0123456789") {
-		s.fname = fmt.Sprintf("ubuntu-base-%s-base-%s.tar.gz",
-			definition.Image.Release, definition.Image.ArchitectureMapped)
-	} else {
-		// if release is non-numerical, find the latest release
-		s.fname = getLatestRelease(definition.Source.URL,
-			definition.Image.Release, definition.Image.ArchitectureMapped)
-		if s.fname == "" {
-			return fmt.Errorf("Couldn't find latest release")
+	patterns := []string{
+		"/lib/*-linux-gnu/ld-linux*.so.2",
+		"/lib/*-linux-gnu/libc.so.6",
+		"/lib/*-linux-gnu/libdl.so.2",
+		"/lib/*-linux-gnu/libfuse.so.2",
+		"/usr/lib/*-linux-gnu/liblz4.so.1",
+		"/lib/*-linux-gnu/liblzma.so.5",
+		"/lib/*-linux-gnu/liblzo2.so.2",
+		"/lib/*-linux-gnu/libpthread.so.0",
+		"/lib/*-linux-gnu/libz.so.1",
+		"/lib/*-linux-gnu/libsquashfuse.so.0",
+		"/lib/*-linux-gnu/libfuseprivate.so.0",
+	}
+
+	for _, p := range patterns {
+		matches, err := filepath.Glob(p)
+		if err != nil {
+			return err
+		}
+
+		if len(matches) != 1 {
+			continue
 		}
+
+		target := filepath.Join(rootfsDir, "lib", filepath.Base(matches[0]))
+
+		err = lxd.FileCopy(matches[0], target)
+		if err != nil {
+			return err
+		}
+
+		err = os.Chmod(target, 0755)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Download init script
+	res, err := http.Get("https://raw.githubusercontent.com/lxc/lxc-ci/master/images/ubuntu-core/init")
```

Comment:
So I'd actually like to get rid of that script as part of moving to distrobuilder, so we should bundle that in distrobuilder.

Reply:
I used [packr](https://github.com/gobuffalo/packr) to bundle the init file as well as the other binaries and libraries. This enables any distro/release to create correct Ubuntu Core images.

Reply:
Hmm, so I don't like that part so much. It comes with some serious problems:
 - How do we handle security updates to those bits?
 - How do we respect the relevant licenses for all of those? We may be required to provide the sources which built those binaries which may be tricky.
 - It's going to be a PITA to add more architectures as we'll need per-series and per-architecture version of all of them.

While certainly not ideal, one option could be to:
 - Download the matching non-core version of our image (from https://images.linuxcontainers.org)
 - Install the relevant packages
 - Copy the binaries from there

So then in the ubuntu core yaml, we'd both pass the core version (e.g. `18`) and the ubuntu version (e.g. `18.04`) so we know what bits to go fetch.

Thoughts?

Reply:
> * Download the matching non-core version of our image (from https://images.linuxcontainers.org)
> * Install the relevant packages


I was trying to avoid having to run a container when wanting to build an image, but yes, that is of course an option.

> So then in the ubuntu core yaml, we'd both pass the core version (e.g. 18) and the ubuntu version (e.g. 18.04) so we know what bits to go fetch.

Do we really need to pass both? Can't we assume that Ubuntu Core XX is based on Ubuntu XX.04 or something?

Reply:
I suppose we can for now, if that becomes a problem, we can change that.

You won't need to run a container, just grab `rootfs.tar.xz`, unpack it, chroot, install what you need, get the binaries out.

Reply:
Done.